### PR TITLE
[BEAM-11747] Narrow list of unsupported types in BeamJavaUdfCalcRule.

### PR DESCRIPTION
**Pull request by ibzib**
All remaining internal compliance test errors and crashes were caused by either DATE, TIME, or DATETIME.

R: @apilloud 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by apilloud** [comment]
With this approach, new types are accepted by default. What is the advantage of rejecting unsupported types rather than allowing supported types?

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java [hunk 1]
```diff
@@ -174,29 +174,20 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
...
-              case BIGINT:
-              case DOUBLE:
-              case BOOLEAN:
-              case VARCHAR:
-              case VARBINARY:
-              case DECIMAL:
-                break;
-              default:
-                // Reject unsupported input ref
+              case BIGINT: // BEAM-11989
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:186]: BigInt was actually in the supported types for InputRef before...

**Comment by ibzib** [comment]
> With this approach, new types are accepted by default. What is the advantage of rejecting unsupported types rather than allowing supported types?

I changed it.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java (hunk 1 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:211]: I noticed that, but I wasn't sure why it is safe to include in input refs but not literals?

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java [hunk 2]
```diff
@@ -174,28 +174,39 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
...
               case BOOLEAN:
-              case CHAR:
-              case BINARY:
+              case VARCHAR:
+              case VARBINARY:
               case DECIMAL:
+              case TIME:
+              case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+              case TIMESTAMP:
+              case ARRAY:
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:185]: I don't think there are ARRAY or ROW literals. I'm pretty sure both become calls to constructors.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java [hunk 3]
```diff
@@ -174,28 +174,39 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
...
-              case BIGINT:
               case DOUBLE:
               case BOOLEAN:
               case VARCHAR:
               case VARBINARY:
               case DECIMAL:
+              case TIME:
+              case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+              case TIMESTAMP:
+              case ARRAY:
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:204]: I believe you need to recurse into `((RexInputRef) rexNode).getType().getComponentType()` to verify the array type is supported.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java (hunk 1 above)
Comment (apilloud) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:207]: Literals and input refs have a completely different implementation. BigInt is represented as a `BigInteger` in InputRef and a `BigDecimal` in Literals. (That is possibly a bug in Calcite.)

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java [hunk 4]
```diff
@@ -174,28 +174,39 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
...
+              case ROW:
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:205]: I believe you need to recurse into `((RexInputRef) rexNode).getType().getFieldList()` to verify the struct only contains supported types.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java [hunk 5]
```diff
@@ -174,28 +174,39 @@ static boolean hasOnlyJavaUdfInProjects(RelOptRuleCall x) {
           } else if (rexNode instanceof RexLiteral) {
             SqlTypeName typeName = ((RexLiteral) rexNode).getTypeName();
             switch (typeName) {
-              case NULL:
+              case DOUBLE:
               case BOOLEAN:
-              case CHAR:
-              case BINARY:
+              case VARCHAR:
```
[the remark this answers is not in the document]
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:179]: I was doing some debugging and found that Calcite treats a null string as `VARCHAR`. So we'll need to include both.

[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java:179]: I was under the impression that null string should be treated as `NULL`, but I might be wrong here.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java (hunk 2 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java]: done

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java (hunk 3 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java]: done

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java (hunk 4 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSQLQueryPlanner.java]: done

**Comment by ibzib** [comment]
@apilloud I addressed your comments. I also did some refactoring because the type matching for the rules should belong in the rules themselves. PTAL

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java [hunk 6]
```diff
@@ -52,4 +117,57 @@ public RelNode convert(RelNode rel) {
...
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Returns true only if the input ref can be correctly implemented by {@link
+   * org.apache.beam.sdk.extensions.sql.impl.rel.BeamCalcRel} in ZetaSQL.
+   */
+  private static boolean udfSupportsInput(RexInputRef rexInputRef) {
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java:147]: nit: I'm not sure this function adds value, consider just calling `udfSupportsInputType` directly?

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java [hunk 7]
```diff
@@ -52,4 +117,57 @@ public RelNode convert(RelNode rel) {
...
+  private static boolean udfSupportsLiteral(RexLiteral rexLiteral) {
+    switch (rexLiteral.getType().getSqlTypeName()) {
+      case BIGINT:
+      case BOOLEAN:
+      case DECIMAL:
+      case DOUBLE:
+      case TIME:
+      case TIMESTAMP:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      case VARBINARY:
```
Comment (apilloud) [inline sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java:134]: I think we had a previous discussion on this CL about `CHAR`, `BINARY`, and `NULL`... They are all missing here.

**Review by apilloud — APPROVED** [review]
LGTM

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java (hunk 7 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java:134]: Oops, missed those. Done.

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java (hunk 6 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java]: Done. (Also changed literals for symmetry.)

File: sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java (hunk 7 above)
Comment (ibzib) [reply sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java:134]: Done. (Also changed literals for symmetry.)

**Review by apilloud — APPROVED** [review]
LGTM

**Comment by ibzib** [comment]
Finally got this to pass compliance tests.

**Pull request merged**